### DISCUSSION TITLE
Enable access to scope for every node in the AST

### DISCUSSION THREAD

**Discussion by goelayu** [swc-project/swc#6576] [Q&A] [answered]
Inspired by the `scope` feature available in babel while traversing the ast, wondering if something similar is already available, or could be added to swc?

Example use of how scope information can aid transformations:
```
visitor {
   Identifier(path){
        if (path.scope.hasOwnBindings(path.node.name) // this identifier is a locally declared variable
            path.node.name = "x"
    }
}
```

Surely something similar can be done by tracking additional state while using swc, but wondering if any API already exists for this purpose, or any plans to add it?

**Comment by kdy1** [comment, marked as answer]
Simply, it can't be added.
If you study Rust more, you will understand the constraint

**Comment by goelayu** [comment]
Thanks for the prompt response! 

In my mini benchmarks, I am seeing swc performing 3x faster than babel so I am very much tempted to migrate my plugins to swc, but the lack of helpers will mean more developer time. 
The unfortunate trade-off between developer time and compute time ugh!!

(Thanks for writing this tool though, much appreciated)
